src/v/cloud_topics/batch_cache/batch_cache.cc

```diff
@@ -52,4 +68,20 @@ batch_cache::get(const model::ntp& ntp, model::offset o) {
     return std::nullopt;
 }
 
+void batch_cache::cleanup_index_entries() {
+    // NOTE: the memory is reclaimed asynchronously.  In some cases
+    // the index may no longer reference any live entries.  If this
+    // is the case we need to delete the batch_cache_index from the
+    // '_index'  collection to avoid accumulating orphaned entries.
+    chunked_vector<model::ntp> empty;
+    for (const auto& entry : _index) {
+        if (entry.second->empty()) {
+            empty.push_back(entry.first);
+        }
+    }
+    for (const auto& ntp : empty) {
+        _index.erase(ntp);
+    }
+}
```

Comment:
Since the members here are chunked vectors and chunked_hash_map this could be large and yield stalls? We could move back to `absl::btree_map` and use the trick used for wasm GC: https://github.com/redpanda-data/redpanda/blob/1308cbd050d48f7ac4be24a835c52df42f3d8ebd/src/v/wasm/cache.cc#L36-L60